<a href="https://colab.research.google.com/github/Aayush360/Natural_langauge_processing/blob/main/Text_Generator_using_LSTM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# that can generate text related to describing hotels for the city of Mumbai

In [6]:
# notebook imports

In [7]:
import nltk
from nltk.corpus import stopwords
import pandas as pd
import numpy as np

import re

from keras.preprocessing.sequence import pad_sequences
from keras.utils import np_utils
from keras.models import Sequential
from keras.layers import Dense,LSTM,Dropout,Embedding

In [8]:
nltk.download('stopwords')


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

### Reading the data 

[Data-Source](https://data.world/promptcloud/hotels-on-makemytrip-com)

In [9]:
data=pd.read_csv('hotel.csv')

In [10]:
data.head()

area  ...                    location
0      Hardasji Ki Magri  ...  POINT(73.667805 24.581788)
1          Near Nai Gaon  ...  POINT(73.644242 24.544342)
2  Near Bagore Ki Haveli  ...  POINT(73.681953 24.579942)
3                  Dabok  ...  POINT(73.855194 24.622607)
4           East Udaipur  ...      POINT(73.6822 24.5801)

[5 rows x 34 columns]

In [11]:
data.city.value_counts()

NewDelhiAndNCR    1163
Goa               1122
Mumbai             543
Jaipur             534
Bangalore          512
                  ... 
Forbesganj           1
ChambaHimachal       1
Ramtek               1
Vengurla             1
Etawah               1
Name: city, Length: 770, dtype: int64

A substantial amount of hotel data form the city of Mumabai is available (543). Lets get description for Mumbai Hotels

In [12]:
array=['Mumbai']
data_mum = data.loc[data['city'].isin(array)]

In [13]:
data_mum.head()

area  ...                             location
294           Charai  ...           POINT(72.974173 19.198414)
309   Andheri (East)  ...           POINT(72.882023 19.117263)
321             Khar  ...             POINT(72.827976 19.0914)
334   Andheri (East)  ...               POINT(72.8431 19.0162)
1238  Andheri (East)  ...  POINT(72.87104239999996 19.1091936)

[5 rows x 34 columns]

Since we are interested in generating hotel description we will only keep the hotel overview column, since other columns will not be required for our analysis

Also let us remove all the descriptions that are empty

In [14]:
data_mum = data_mum.hotel_overview
data_mum = data_mum.dropna()

In [15]:
data_mum.head() # we have series of reviwes about Mumbai hotels

294     Nestled in Mumbai, a city with strong historic...
309     3 km from Chhatrapati Shivaji International Ai...
321     Location Hotel Royal Garden is situated on Juh...
334     City Guest House is a beautiful property locat...
1238    Sai Residency Hotel is situated in the City of...
Name: hotel_overview, dtype: object

### Pre-Process the data

In [16]:
# perfrom case-folding (converting to lowercase, stopword removal, keeping only alphabetic data), Also discard single character word

In [17]:
stop = set(stopwords.words('english'))

def stopword_removal(data_point):
  data= [x for x in data_point.split() if x not in stop]
  return data


In [18]:
ex = 'Nestled in Mumbai, a city with strong historic'
ex = stopword_removal(ex)
print(ex)

['Nestled', 'Mumbai,', 'city', 'strong', 'historic']


In [19]:
# method for overall data cleansing

In [20]:
def clean_data(data): 
  '''works for single sentence and returns cleaned tokens of words in the sentence along with unique_words'''
  cleaned_data =[]
  all_unique_words_in_each_description=[]
  for entry in data:
    entry = re.sub(pattern='[^a-zA-Z]',repl=" ", string=entry) # check if alphabet, remove non-alphabet
    entry = re.sub(r'\b\w{0,1}\b', repl=" ", string=entry)  # remove single word or blank character/word
    entry =entry.lower()
    entry = stopword_removal(entry)
    cleaned_data.append(entry) # list with in list 0(1) time-complexity
    unique = list(set(entry))
    all_unique_words_in_each_description.extend(unique) # appended list O(K) time-complexity, k- lenght of list need to be added
  return cleaned_data,all_unique_words_in_each_description

In [21]:
# function to make vocabulary

def unique_words(data):
  unique_words = set(data)
  return unique_words, len(unique_words)

#### Applying data cleaning to our df_mum

In [22]:
cleaned_data, all_unique_words_in_each_description = clean_data(data_mum)

In [23]:
unique_words, len_unique_words = unique_words(all_unique_words_in_each_description)

In [24]:
len_unique_words  # this is our vocab_size

3395

In [25]:
## peeking at one of the cleaned data

len(cleaned_data[0])

136

### Build Mapping for Words

### word --> index and index --> word

In [26]:
def build_indices(unique_words):
  word_to_index = {}
  index_to_word = {}
  for i,word in enumerate(unique_words):
    word_to_index[word]=i
    index_to_word[i]=word
  return word_to_index, index_to_word

In [27]:
word_to_index, index_to_word = build_indices(unique_words)

### Prepare Training Corpus

In [28]:
def prepare_corpus(corpus,word_to_index):
  sequences= []
  for line in corpus:
    tokens = line
    for i in range(1, len(tokens)):
      i_gram_sequence = tokens[:i+1]
      i_gram_sequence_ids = []

      for j,token in enumerate(i_gram_sequence):
        i_gram_sequence_ids.append(word_to_index[token])
      sequences.append(i_gram_sequence_ids)
  return sequences

In [29]:
sequences=prepare_corpus(cleaned_data,word_to_index)
max_sequence_len = max([len(x) for x in sequences]) # find sentence with maximum length

In [30]:
len(sequences) # our sequences contains 51836 list with colleciton of n-grams for each sentence staritng from length 2

51836

In [31]:
max_sequence_len # so the max len of sentece after cleaning and tokenizing is 308

308

In [32]:
sequences[0] # looking at the first sentence and its i-gram indices (first sentence with 2 words indexes)

[285, 2838]

In [33]:
sequences[1]

[285, 2838, 474]

In [34]:
# let us see which words are mapped to these indices

print(index_to_word[3125])
print(index_to_word[66])
print(index_to_word[1864])

delved
delicacies
around


In [35]:
# Finally, we have correctly build our sequences

In [36]:
## Preparing the data for trainng the model
# we should pad the input to the longest words in the sequences 308 in this case
# also we need to split the data into dependent and independent variable

In [37]:
def build_input_data(sequences,max_sequence_len, len_unique_words):
  sequences = np.array(pad_sequences(sequences,maxlen=max_sequence_len, padding='pre'))
  X = sequences[:,:-1]
  y = sequences[:,-1]
  y = np_utils.to_categorical(y,len_unique_words) # convert correspoding row index to 1, and len of each word is vector of length len_unique_words
  return X,y


In [38]:
X,y = build_input_data(sequences,max_sequence_len,len_unique_words)

### Building The model

input_dim: This is the size of the vocabulary in the text data. For example, if your data is integer encoded to values between 0-10, then the size of the vocabulary would be 11 words.
output_dim: This is the size of the vector space in which words will be embedded. It defines the size of the output vectors from this layer for each word. For example, it could be 32 or 100 or even larger. Test different values for your problem.
input_length: This is the length of input sequences, as you would define for any input layer of a Keras model. For example, if all of your input documents are comprised of 1000 words, this would be 1000.

In [39]:
# 10 indicates that we want a Dense embedding of size 10 as output from our model.
# 128 LSTM units in the hidden layer, dimensionality of the inner cells in the LSTM layer.
# randomly drop off 20% of neurons from the network using the Dropout layer.
#  multi-class classification problem, and so the softmax activation function is used.

In [40]:
def create_model(max_sequence_len, len_unique_words):
  model = Sequential()
  model.add(Embedding(len_unique_words,10,input_length=max_sequence_len-1))
  model.add(LSTM(128))
  model.add(Dropout(0.2))
  model.add(Dense(len_unique_words, activation='softmax'))
  model.compile(loss='categorical_crossentropy',optimizer='adam')
  return model


In [41]:
model = create_model(max_sequence_len, len_unique_words)
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding (Embedding)        (None, 307, 10)           33950     
_________________________________________________________________
lstm (LSTM)                  (None, 128)               71168     
_________________________________________________________________
dropout (Dropout)            (None, 128)               0         
_________________________________________________________________
dense (Dense)                (None, 3395)              437955    
Total params: 543,073
Trainable params: 543,073
Non-trainable params: 0
_________________________________________________________________


In [ ]:
model.fit(X,y,batch_size=512, epochs=100)

### Testing and Text Generation

In [43]:
def generate_text(seed_text, next_words, model, max_seq_len):
  for _ in range(next_words):
    cleaned_data = clean_data([seed_text])
    sequences = prepare_corpus(cleaned_data[0],word_to_index)
    sequences = pad_sequences([sequences[-1]],maxlen=max_seq_len-1,padding='pre')
    predicted = model.predict_classes(sequences, verbose=0)
    output_word=''
    output_word = index_to_word[predicted[0]]
    seed_text = seed_text+" "+output_word
  return seed_text.title()

In [44]:
print(generate_text("mumbai is a city of beautiful",20,model,max_sequence_len))


/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/engine/sequential.py:450: UserWarning: `model.predict_classes()` is deprecated and will be removed after 2021-01-01. Please use instead:* `np.argmax(model.predict(x), axis=-1)`,   if your model does multi-class classification   (e.g. if it uses a `softmax` last-layer activation).* `(model.predict(x) > 0.5).astype("int32")`,   if your model does binary classification   (e.g. if it uses a `sigmoid` last-layer activation).
  warnings.warn('`model.predict_classes()` is deprecated and '


Mumbai Is A City Of Beautiful Residency Located Distance Km Chhatrapati Shivaji International Airport Km Chhatrapati Shivaji International Airport Km Chhatrapati Shivaji International Airport Km Chhatrapati


Hyperparameter tuning, along with building more complex and larger models, can help in generating better results.